Gutenberg: Fixes for Gutenberg v9.4.1

fullofcaffeine opened the pull request:
### TODO

- [x] Fix tests while keeping them compatible with GB < 9.4

This also includes some additional improvements as per the discussion [here](https://github.com/Automattic/wp-calypso/pull/47723#discussion_r530995080).

fullofcaffeine committed:
Fix XPath to the "Code editor" options button, it now wraps its text in a span

fullofcaffeine committed:
Fix selector hierarchy for the block button in the main inserter button

Use a selector that works for GB <= 9.4.

fullofcaffeine commented on test/e2e/lib/gutenberg/gutenberg-editor-component.js:
```diff
@@ -353,7 +353,7 @@ export default class GutenbergEditorComponent extends AsyncBaseContainer {
 
 		const inserterBlockItemSelector = By.css(
-			`.edit-post-layout__inserter-panel .block-editor-inserter__block-list button.editor-block-list-item-${ prefix }${ blockClass }`
+			`.edit-post-layout__inserter-panel .block-editor-block-types-list button.editor-block-list-item-${ prefix }${ blockClass }`
```
Probably related to the change that led to this failure, the search results no longer seem to show the block category (it also hides the tabs). See the example of the "Dynamic HR" block:

| GB 9.3 | GB 9.4 RC1 |
| ------------- | ------------- |
| ![gb93](https://user-images.githubusercontent.com/81248/99472325-fd8b2380-290d-11eb-9e71-7090c5cd1ff2.png) | ![gb94](https://user-images.githubusercontent.com/81248/99472331-00861400-290e-11eb-99b0-3b93d442c039.png) |

I'm pretty sure this is just a UX change and not a bug, but taking a note here in case it causes confusion down the road.

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@fullofcaffeine please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/954539). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

fullofcaffeine committed:
Use `getAttribute("innerText")` because `getText()` is unstable

`getText()` was returning an empty string in my local env.

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@fullofcaffeine please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/956080). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

fullofcaffeine committed:
Fix tests that check for Earn and Grow blocks

They were checking for the categories, but these are not shown anymore. We now check that the blocks are present after the inserter search for the category keyword.

fullofcaffeine commented on test/e2e/specs/wp-calypso-gutenberg-post-editor-spec.js:
```diff
@@ -281,22 +281,43 @@ describe( `[${ host }] Calypso Gutenberg Editor: Posts (${ screenSize })`, funct
+
+				// With deepEqual we're also testing the sort order they're shown in the
+				// inserter. If that's not desirable, we can use the sameMembers matcher.
+				assert.deepEqual( await gEditorComponent.getShownBlockInserterItems(), [
```
Before it was just checking if the actual DOM element for the category tab was present. Category is not rendered as part of the inserter search UI anymore, though (see https://github.com/Automattic/wp-calypso/pull/47533/files#r525642380). 

We can either:

1. Check for the actual blocks. This is what I did for now;
2. Get rid of those tests?

Also, we I decided to use `deepEqual` in order to get sort order testing for free. If that's overkill, we can use `sameMembers`.

Thoughts?

fullofcaffeine committed:
Add inserterBlockItemSelector fallback for GB < 9.4

fullofcaffeine committed:
Make Earn and Grow cat/block tests shallow in order to make them compatible with GB < 9.4

fullofcaffeine commented on test/e2e/specs/wp-calypso-gutenberg-post-editor-spec.js:
```diff
@@ -281,22 +281,49 @@ describe( `[${ host }] Calypso Gutenberg Editor: Posts (${ screenSize })`, funct
-					await gEditorComponent.isBlockCategoryPresent( 'Earn' ),
-					true,
-					'Earn (Jetpack) blocks are not present'
+					[
```
> Before it was just checking if the actual DOM element for the category tab was present. Category is not rendered as part of the inserter search UI anymore, though (see https://github.com/Automattic/wp-calypso/pull/47533/files#r525642380).
> 
> We can either:
> 
>     Check for the actual blocks. This is what I did for now;
>     Get rid of those tests?
> 
> Also, we I decided to use deepEqual in order to get sort order testing for free. If that's overkill, we can use sameMembers.
> 
> Thoughts?

I momentarily forgot that this should be compatible with GB < 9.4 as well. I initially thought about having a condition that tried to find the tab element unique to 9.3, then run the previous assertion, and if not, run the one for 9.4. I decided to simplify further and make it a more shallow check, that actually doesn't rely on the order and only checks that the elements are present.

This means it won't test the sorting order and will not fail if additional blocks are shown (which happens on 9..3, see the screenshot comparison below). This works for < 9.4 and 9.4. I think it also better satisfies the test description, too. Let me know your thoughts.

| GB 9.3 | GB 9.4 |
| ------------- | ------------- |
| ![gb93](https://user-images.githubusercontent.com/81248/99751538-fef05380-2aa7-11eb-964d-30c2a0035127.png) | ![gb94](https://user-images.githubusercontent.com/81248/99751542-01eb4400-2aa8-11eb-9a67-b96360412d6d.png) |

fullofcaffeine replied:
Also, I couldn't find an assertion method in `assert` that would assert that a array included a subset, so I had to [improvise](https://github.com/Automattic/wp-calypso/pull/47533/files#diff-2d3083bcc78e9ea918bc574151a50d2e126be3ee69611241fa98945bf864c603R323). Let me know if I missed something. cc @bsessions85

ockham replied:
Are we using Jest to run these? 'cause Jest has [`expect.arrayContainingArray`](https://jestjs.io/docs/en/expect#expectarraycontainingarray).

If we're _not_ using Jest, maybe we should consider migrating to it (as a follow-up)? I think Gutenberg is using it for its e2e tests, no?

fullofcaffeine replied:
We're using Mocha. I thought we were using Chai too, but it appears it's not the case. I think migrating to Jest makes sense as a follow up issue. I'd like to know @bsessions85's thoughts on that, too.

bsessions85 replied:
I've looked in to moving to jest, but the way our tests are set up, I couldn't get the the full "describe" block to re-run, just the failed steps, which didn't help. I could get close by hacking some packages, but it was going to be a maintenance problem. I'll look in to it again as its been about a year. We could look at using Chai with Mocha as well for now, even if it is just in this file to start

ockham replied:
Cycling back to this:

> I momentarily forgot that this should be compatible with GB < 9.4 as well.

Does it, though? If we land this PR briefly after enabling 9.4 in production, we don't necessarily need backwards-compat, do we? Or is that related to the block upgrade/migration tests?

fullofcaffeine replied:
Ideally, but not strictly necessary. We can indeed land it after we ship 9.4.

fullofcaffeine committed:
Make test aware of the fact Contact Info doesn't give focus to its inner child upon init anymore

fullofcaffeine commented:
Applied the `[Status] Needs e2e Testing` tag in order to run the E2E specs on non-edge sites (GB 9.3) to see if the changes are indeed 9.3-compatible. For the record, all checks were passing 🟢 for edge (`[Status] Needs e2e Testing Gutenberg Edge` tag).

fullofcaffeine commented on test/e2e/lib/gutenberg/gutenberg-editor-component.js:
```diff
@@ -273,11 +275,26 @@ export default class GutenbergEditorComponent extends AsyncBaseContainer {
 		title = title.charAt( 0 ).toUpperCase() + title.slice( 1 ); // Capitalize block name
 		let blockClass = kebabCase( title.toLowerCase() );
-		let hasChildBlocks = false;
+		const initsWithChildFocus = false;
```
Note to self: This should not be a `const`.

fullofcaffeine committed:
Simplify XPath for selecting the switch to code editor button

fullofcaffeine committed:
Simplify addBlock by extracting the block selector switch logic to a separate function

Paving the way to refactoring it later to use the actual block classes (`lib/gutenberg/blocks`) by moving the settings data there, too.

Also added the proper jsdoc.

fullofcaffeine committed:
Run assertion per block to make it easier to pinpoint the root of the failure

fullofcaffeine committed:
Simplify by using webdriver.promise.map instead of regular map wrapped by Promise.all

fullofcaffeine committed:
Move ariaLabel definition logic to getBlockSelectorSettings and add back the default prefix attribue

fullofcaffeine committed:
fixup! Fix webdriver import

fullofcaffeine committed:
Revert "Simplify XPath for selecting the switch to code editor button"

This reverts commit c33f574ba68c5d4c5a3eea4722d99407b3a80731.

fullofcaffeine committed:
fixup! Fix jetpack subscriptions prefix

fullofcaffeine commented on test/e2e/lib/gutenberg/gutenberg-editor-component.js:
```diff
@@ -273,100 +276,129 @@ export default class GutenbergEditorComponent extends AsyncBaseContainer {
+	 * @typedef {Object} BlockSelectorSettings
+	 * @property {string} title The block title as it appears in the inserter
+	 * @property {string} blockClass The suffix that's part of a wrapper CSS class that's used to select the block button in the inserter.
+	 * Calcualted from the title if not present.
```
Calcualted > Calculated. Will fix later in order to not start yet another build now.

ockham commented on test/e2e/lib/gutenberg/gutenberg-editor-component.js:
```diff
@@ -273,100 +276,129 @@ export default class GutenbergEditorComponent extends AsyncBaseContainer {
+	 * @property {string} title The block title as it appears in the inserter
+	 * @property {string} blockClass The suffix that's part of a wrapper CSS class that's used to select the block button in the inserter.
+	 * Calcualted from the title if not present.
+	 * @property {string} prefix Also used to build the CSS class that's used t o select the block in the inserter.
```
```suggestion
	 * @property {string} prefix Also used to build the CSS class that's used to select the block in the inserter.
```

ockham commented:
General question -- potentially for a separate PR: Rather than maintaining our own [`addBlock` implementation](https://github.com/Automattic/wp-calypso/pull/47533/files#diff-b7f8735b885b065dc18c161fbf8de3c5bfecafd5fd04319842a8eea5f5c32180R382), could we use [`insertBlock`](https://github.com/WordPress/gutenberg/blob/7b5d1a7a37088feab8d36fcce4c077579f90588c/packages/e2e-test-utils/src/inserter.js#L124-L138) from `@wordpress/e2e-test-utils` instead? Seems like that should save us a bunch of work, no?

ockham commented on test/e2e/lib/gutenberg/gutenberg-editor-component.js:
```diff
@@ -273,100 +276,129 @@ export default class GutenbergEditorComponent extends AsyncBaseContainer {
 		if ( title === 'Heading' ) {
-			const deprecatedSelector = insertedBlockSelector.value.replace(
-				selectorAriaLabel,
-				'Write heading…'
-			);
+			const deprecatedSelector = insertedBlockSelector.value.replace( ariaLabel, 'Write heading…' );
 			insertedBlockSelector = By.css( `${ insertedBlockSelector.value }, ${ deprecatedSelector }` );
 		}
```
Seems like we can remove this altogether?

See https://github.com/Automattic/wp-calypso/pull/47533/files#diff-b7f8735b885b065dc18c161fbf8de3c5bfecafd5fd04319842a8eea5f5c32180L364

Edit: Oh, or rather just remove the `title === 'Heading'` check maybe? I might've misread at first.

fullofcaffeine replied:
We can remove the whole block of code. The deprecated selector was only needed for GB < 9.1. The condition there is to make sure it applies only when inserting the `Heading` block Thanks for pointing this out :)

ockham commented:
> General question -- potentially for a separate PR: Rather than maintaining our own [`addBlock` implementation](https://github.com/Automattic/wp-calypso/pull/47533/files#diff-b7f8735b885b065dc18c161fbf8de3c5bfecafd5fd04319842a8eea5f5c32180R382), could we use [`insertBlock`](https://github.com/WordPress/gutenberg/blob/7b5d1a7a37088feab8d36fcce4c077579f90588c/packages/e2e-test-utils/src/inserter.js#L124-L138) from `@wordpress/e2e-test-utils` instead? Seems like that should save us a bunch of work, no?

Ah, that doesn't return the ID of the inserted block. Well we might still consider adding that feature to `@wordpress/e2e-test-utils`, as it seems like a useful general-purpose e2e utility.

fullofcaffeine committed:
Remove deprecated heading selector logic

fullofcaffeine committed:
Fix spelling in comment
